arxiv/integration/api/service.py

```diff
@@ -0,0 +1,261 @@
+        elif resp.status_code >= status.HTTP_400_BAD_REQUEST:
+            raise BadRequest(f'Bad request: {resp.content}', resp)
+        elif resp.status_code not in expected_code:
+            raise RequestFailed(f'Unexpected code: {resp.status_code}', resp)
+
+    def _parse_location(self, location: str) -> str:
+        endpoint_path = urlparse(self._endpoint).path
+        if self._endpoint in location:
+            _, location = location.split(self._endpoint, 1)
+        elif endpoint_path in location:
+            _, location = location.split(endpoint_path, 1)
+        return location
+
+    def request(self, method: str, path: str, token: Optional[str] = None,
+                expected_code: List[int] = [status.HTTP_200_OK],
+                allow_2xx_redirects: bool = True,
+                **kwargs) -> requests.Response:
+        """Make an HTTP request with error/code handling."""
+        if token is not None:
+            if 'headers' not in kwargs:
+                kwargs['headers'] = {}
+            kwargs['headers'].update({'Authorization': token})
+        try:
+            resp = getattr(self._session, method)(self._path(path), **kwargs)
+            logger.debug('Got response %s', resp)
+        except requests.exceptions.SSLError as e:
+            raise SecurityException('SSL failed: %s' % e) from e
+        except requests.exceptions.ConnectionError as e:
+            raise ConnectionFailed('Could not connect: %s' % e) from e
+
+        # 200-series redirects are a common pattern on our projects, so these
+        # should be supported.
+        is_2xx = status.HTTP_200_OK < resp.status_code \
+            and resp.status_code < status.HTTP_300_MULTIPLE_CHOICES
+        if allow_2xx_redirects and is_2xx and 'Location' in resp.headers:
+            loc = self._parse_location(resp.headers['Location'])
+            logger.debug('Following 2xx redirect to %s', loc)
+            resp = self.request('get', loc, token, expected_code,
+                                allow_2xx_redirects)
+
+        self._check_status(resp, expected_code)
+        return resp
+
+    def json(self, method: str, path: str, token: Optional[str] = None,
+             expected_code: List[int] = [status.HTTP_200_OK], **kwargs) \
+            -> Tuple[dict, int, MutableMapping]:
+        """
+        Perform an HTTP request to a JSON endpoint, and handle any exceptions.
+
+        Returns
+        -------
+        dict
+            Response content.
+        int
+            HTTP status code.
+        dict
+            Response headers.
+
+        """
+        response = self.request(method, path, token, expected_code, **kwargs)
+        try:
+            return response.json(), response.status_code, response.headers
+        except json.decoder.JSONDecodeError as e:
+            raise BadResponse(f'Could not decode', response) from e
+
+    def get_status(self) -> dict:
+        """Get the status of the file management service."""
```

Comment:
This probably doesn't need "file management"

Reply:
oh yuck; fixed